### ISSUE TITLE
[UX] Automatic menu link creation for nodes.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#901]
Use case: Force nodes of a specific content type to have links created under a specific menu (without allowing the end user to change parent menu or to not create a menu link). For example, each time a content author creates a new "Service" or "Product" node, there should **always** be a respective menu link created under the "Services"/"Products" menu items.

As it is now, this is not possible. The user has to remember to select "Provide a menu link" checkbox. [Default Menu Link](https://www.drupal.org/project/default_menu_link) does this, but still does not handle menu parent and menu link title.

Also, there are cases where you want to **force** the menu link creation. Even with [Default Menu Link](https://www.drupal.org/project/default_menu_link) that sets the checkbox to enabled by default, the user can still disable it even if the site admin means them not to. Modules like [Menu Force](https://www.drupal.org/project/menu_force) come to play then.

Lastly, if you force menu creation, you need a way to also force menu link title creation/update (à la pathauto and/or pathauto_persist). [Auto Menu Title](https://www.drupal.org/project/auto_menutitle) to the rescue!

All these are tiny modules that count a little over 1.000 installations each, but I'm sure that these features are not uncommon. I guess others that need these resort to `form_alter` and the likes in custom code. Can we please consider these features for core?

**Comment by klonos** [comment]
...I was thinking perhaps now with token in core, allow the menu link title field to accept tokens?

**Comment by klonos** [comment]
The [NodeAutoMenu](https://www.drupal.org/sandbox/wojtha/1331132) sandbox (a D7 fork of [Auto Menu](https://www.drupal.org/project/automenu)) does all this in a very nice way I think:

![nodeautomenu-ui](https://cloud.githubusercontent.com/assets/2423362/7441096/a4ac1650-f11b-11e4-803f-e61a83574a05.png)

The "Default the "Provide a menu link" item to checked." checkbox in the screenshot is because I have [Default Menu Link](https://www.drupal.org/project/default_menu_link) installed.

The "Nodeautomenu handler" section is because NodeAutoMenu comes with a submodule that allows menu placement based on taxonomy. I don't think this is needed, but if someone sees a common use case for it, please let us know.

Also the "Chop first/last ancestor" options, I don't personally see a use for, but could be wrong.

So basically, an option to automatically create/update menu links and another to hide this setting from users (so to force the autocreation instead of give the option to) should be enough and at the same time not complicate the UI too much.

As I said, now that Token is going to be in core, perhaps we should add a field to allow tokens to be used for the node title and have it default to ```node:title```.

Here's how the node edit form looks when the "Hide menu settings" is enabled for the content type:

![nodeautomenu-node_edit_form](https://cloud.githubusercontent.com/assets/2423362/7441125/a4f97b96-f11d-11e4-8da1-87d2d8fe175b.png)

**Comment by docwilmot** [comment]
Wouldnt that be better off in contrib though? Not a terribly common requirement, I'd think.

**Comment by klonos** [comment]
It is relatively common for most of my customers' sites. They keep coming back to me with the request to either have the option checked by default (because they forget to enable it the first time they create a piece of content and then they have to re-edit the same content just to add it to a menu) or because they find content authors having placed the menu links in the wrong (sub)menu and it messes their menu structure. Both use cases make the site owner believe that the site/Drupal/Backdrop "is not working right".

Think of this feature as the extension of what Pathauto does but for menus. "Menuauto" if you like.

**Comment by BWPanda** [comment]
I like this idea. I always use [Auto Menu Title](https://www.drupal.org/project/auto_menutitle) on my Page content type to make menu items by default and [Simplify](https://www.drupal.org/project/simplify) to prevent users undoing this.

Here's what I recommend (for either core or a contrib module, whatever's decided):

- A checkbox to set the default value of 'Provide a menu link' on node edit forms
- A textfield to allow tokens to make up the default menu link title
- Checkboxes to show/hide fields on the node edit form

![firefox_screenshot_2017-05-03t06-46-14 842z](https://cloud.githubusercontent.com/assets/2385329/25650448/595b3cbc-3020-11e7-91bd-906ee3d239c0.png)

I believe this combination of options should allow for everything discussed above...

**Comment by BWPanda** [comment]
I've started a contrib module with this functionality here: https://github.com/backdrop-contrib/auto_menu_settings

Nearly working. A few things todo that I'll add in the next day or two.

**Comment by BWPanda** [comment]
@klonos My module is ready now, so feel free to give it a test drive if you can. I'd love for some feedback before I make the initial release...

**Comment by olafgrabienski** [comment]
@BWPanda Thanks for the module, sounds useful! I've installed it on a test site and played a bit with the settings - not thoroughly tested, but at first sight it looks good!

Only issue so far: there are still some help texts using the term "node" in the menu settings tab. Can you replace them with "content" equivalents? Here's a list with the "node" terms:

- `Set the default value of the 'Provide a menu link' checkbox on the node form.`
- `Set the default value of the 'Menu link title' field on the node form. This field supports tokens.`
- `Show these fields to the node author. Untick fields to hide them.`

**Comment by BWPanda** [comment]
Thanks for the feedback @olafgrabienski! Here's the commit with those changes: https://github.com/backdrop-contrib/auto_menu_settings/commit/59162385476ed9f2865c72b78d12adcf355447ae

**Comment by BWPanda** [comment]
Do we close this in favour of using the [Auto Menu Settings issue queue](https://github.com/backdrop-contrib/auto_menu_settings/issues), or keep it open to discuss adding this functionality to core?

**Comment by olafgrabienski** [comment]
Personally, I'd tend to close it in favor of the contrib queue but it would be good to get new feedback of @klonos and @docwilmot. (In my opinion, the menu settings based on the module look too complex for having them in core all the time.)

**Comment by klonos** [comment]
Thanks @BWPanda for putting the contrib module together 👍The use of tokens to provide a menu title is really awesome!! I cannot wait until I can put this into use next time this problem comes up 😄

...I have taken a quick look and I see some UX issues with the current implementation (some are core issues, but the many options that this module offers makes them stick out).

I agree with @olafgrabienski that too many options over-complicate things and this does not feel like core. I would like to flesh things out in contrib, and then perhaps return here to introduce a very minimal set of features. What I was going for here initially was some way that site admins/builders can use to "lock" their content types to *alway* create a menu item *all* the time, with specific settings. Exposing some/all of these settings to the end users (content authors) was not in scope when I was filing this ticket.

I promise to test this module more and provide feedback by means of ticket to its GitHub queue.

Thanks once again @BWPanda 👍

**Comment by stpaultim** [comment]
@klonos - Is this issue still relevant after - [UX] Allow content to be placed into the menu by default [#1610]

**Comment by klonos** [comment]
Sorry for the late response @stpaultim ...with #1610 we've allowed _defaults_ to be set for menu options when creating content. What we are missing is a way to _force_ these defaults. As soon as you give a user role the "Administer menus and menu items" permission, they are able to change the defaults specified, which in some cases defeats the purpose.

I would argue that even if we hid the options from the node edit form, people with this permission would still be able to edit the menu from the menu edit form. I guess we need #384 then.

**Comment by jenlampton** [comment]
> What we are missing is a way to force these defaults.

I would argue that enforcing any of the defaults set on the content creation form should be left to a contrib project. Default values are always just a starting point.

I'm also not sure this request is an 80% use-case. 

I agree we need https://github.com/backdrop/backdrop-issues/issues/384, but I think this request was adequately handled in https://github.com/backdrop/backdrop-issues/issues/1610. At the very least, we should create a new issue that specifically describes the need, as this one reads as a duplicate of 1610 (or rather, 1610 is a duplicate of this one)
